Data race in pkg/kubelet/config.watchFileChanged

**What happened**:
Unit test failed with a data race in a seemingly unrelated PR https://github.com/kubernetes/kubernetes/pull/63284
Logs: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/63284/pull-kubernetes-bazel-test/47917/
```
==================
WARNING: DATA RACE
Write at 0x00c42051c3c0 by goroutine 97:
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged.func1.3()
      pkg/kubelet/config/file_linux_test.go:341 +0x26a

Previous read at 0x00c42051c3c0 by goroutine 106:
  reflect.typedmemmove()
      GOROOT/src/runtime/mbarrier.go:265 +0x0
  reflect.packEface()
      GOROOT/src/reflect/value.go:119 +0x101
  reflect.valueInterface()
      GOROOT/src/reflect/value.go:978 +0x16a
  reflect.Value.Interface()
      GOROOT/src/reflect/value.go:948 +0x51
  k8s.io/apimachinery/third_party/forked/golang/reflect.Equalities.deepValueEqual()
      vendor/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:221 +0x517
  k8s.io/apimachinery/third_party/forked/golang/reflect.Equalities.deepValueEqual()
      vendor/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:186 +0x9da
  k8s.io/apimachinery/third_party/forked/golang/reflect.Equalities.deepValueEqual()
      vendor/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:172 +0xd7e
  k8s.io/apimachinery/third_party/forked/golang/reflect.Equalities.deepValueEqual()
      vendor/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:186 +0x9da
  k8s.io/apimachinery/third_party/forked/golang/reflect.Equalities.deepValueEqual()
      vendor/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:186 +0x9da
  k8s.io/apimachinery/third_party/forked/golang/reflect.Equalities.deepValueEqual()
      vendor/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:183 +0x14d7
  k8s.io/apimachinery/third_party/forked/golang/reflect.Equalities.deepValueEqual()
      vendor/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:172 +0xd7e
  k8s.io/apimachinery/third_party/forked/golang/reflect.Equalities.deepValueEqual()
      vendor/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:186 +0x9da
  k8s.io/apimachinery/third_party/forked/golang/reflect.Equalities.DeepEqual()
      vendor/k8s.io/apimachinery/third_party/forked/golang/reflect/deep_equal.go:243 +0x2d5
  k8s.io/kubernetes/pkg/kubelet/config.expectUpdate()
      pkg/kubelet/config/file_linux_test.go:400 +0x669
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged.func1()
      pkg/kubelet/config/file_linux_test.go:352 +0x7b0
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged()
      pkg/kubelet/config/file_linux_test.go:363 +0xe2
  k8s.io/kubernetes/pkg/kubelet/config.TestWatchFileChanged()
      pkg/kubelet/config/file_linux_test.go:129 +0x96
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d

Goroutine 97 (running) created at:
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged.func1()
      pkg/kubelet/config/file_linux_test.go:350 +0x771
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged()
      pkg/kubelet/config/file_linux_test.go:363 +0xe2
  k8s.io/kubernetes/pkg/kubelet/config.TestWatchFileChanged()
      pkg/kubelet/config/file_linux_test.go:129 +0x96
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d

Goroutine 106 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d
  testing.runTests()
      GOROOT/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:978 +0x2cd
  main.main()
      bazel-out/k8-fastbuild/bin/pkg/kubelet/config/linux_amd64_race_stripped/go_default_test~/testmain.go:179 +0x3ad
==================
==================
WARNING: DATA RACE
Read at 0x00c420726aa0 by goroutine 97:
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged.func1.3()
      pkg/kubelet/config/file_linux_test.go:343 +0x2e7

Previous write at 0x00c420726aa0 by goroutine 106:
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged.func1()
      pkg/kubelet/config/file_linux_test.go:356 +0x80d
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged()
      pkg/kubelet/config/file_linux_test.go:363 +0xe2
  k8s.io/kubernetes/pkg/kubelet/config.TestWatchFileChanged()
      pkg/kubelet/config/file_linux_test.go:129 +0x96
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d

Goroutine 97 (running) created at:
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged.func1()
      pkg/kubelet/config/file_linux_test.go:350 +0x771
  k8s.io/kubernetes/pkg/kubelet/config.watchFileChanged()
      pkg/kubelet/config/file_linux_test.go:363 +0xe2
  k8s.io/kubernetes/pkg/kubelet/config.TestWatchFileChanged()
      pkg/kubelet/config/file_linux_test.go:129 +0x96
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d

Goroutine 106 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d
  testing.runTests()
      GOROOT/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:978 +0x2cd
  main.main()
      bazel-out/k8-fastbuild/bin/pkg/kubelet/config/linux_amd64_race_stripped/go_default_test~/testmain.go:179 +0x3ad
==================
E0602 12:54:50.094795      15 file_linux.go:61] Unable to read config path "/tmp/dir-test733403562": error while processing inotify event ("/tmp/dir-test733403562": 0x400 == IN_DELETE_SELF): the watched path is deleted
--- FAIL: TestWatchFileChanged (0.88s)
	file_linux_test.go:417: expected empty update, got types.PodUpdate{Pods:[]*v1.Pod{(*v1.Pod)(0xc420575c00)}, Op:0, Source:"file"}
	file_linux_test.go:372: unable to remove test file /tmp/dir-test733403562/test_pod_manifest_0_ch: remove /tmp/dir-test733403562/test_pod_manifest_0_ch: no such file or directory
	testing.go:730: race detected during execution of test
FAIL
```

I couldn't find an existing report about the issue so here it is.

/kind bug
/sig cli

--- a/pkg/kubelet/config/file_linux_test.go
+++ b/pkg/kubelet/config/file_linux_test.go
@@ -321,10 +321,6 @@ func watchFileChanged(watchDir bool, symlink bool, t *testing.T) {
 
 			if watchDir {
 				NewSourceFile(dirName, hostname, 100*time.Millisecond, ch)
-				defer func() {
-					// Remove the file
-					deleteFile(dirName, fileName, ch, t)
-				}()
 			} else {
 				NewSourceFile(file, hostname, 100*time.Millisecond, ch)
 			}
@@ -361,22 +357,6 @@ func watchFileChanged(watchDir bool, symlink bool, t *testing.T) {
 			}
 		}()
 	}
-}
-
-func deleteFile(dir, file string, ch chan interface{}, t *testing.T) {
-	path := filepath.Join(dir, file)
-	if _, err := os.Stat(path); err != nil {
-		// The file might absent if it's renamed, deleted..
-		return
-	}
-	go func() {
-		err := os.Remove(path)
-		if err != nil {
-			t.Errorf("unable to remove test file %s: %s", path, err)
-		}
-	}()
-
-	expectEmptyUpdate(t, ch)
 }
 
 func expectUpdate(t *testing.T, ch chan interface{}, testCase *testCase) {

--- a/pkg/kubelet/config/file_linux_test.go
+++ b/pkg/kubelet/config/file_linux_test.go
@@ -131,6 +131,7 @@ func TestWatchFileChanged(t *testing.T) {
 }
 
 type testCase struct {
+	lock       *sync.Mutex
 	desc       string
 	linkedFile string
 	pod        runtime.Object
@@ -141,6 +142,7 @@ func getTestCases(hostname types.NodeName) []*testCase {
 	grace := int64(30)
 	return []*testCase{
 		{
+			lock: &sync.Mutex{},
 			desc: "Simple pod",
 			pod: &v1.Pod{
 				TypeMeta: metav1.TypeMeta{
@@ -304,11 +306,10 @@ func watchFileChanged(watchDir bool, symlink bool, t *testing.T) {
 			}
 
 			var file string
-			lock := &sync.Mutex{}
 			ch := make(chan interface{})
 			func() {
-				lock.Lock()
-				defer lock.Unlock()
+				testCase.lock.Lock()
+				defer testCase.lock.Unlock()
 
 				if symlink {
 					file = testCase.writeToFile(linkedDirName, fileName, t)
@@ -332,8 +333,8 @@ func watchFileChanged(watchDir bool, symlink bool, t *testing.T) {
 
 			changeFile := func() {
 				// Edit the file content
-				lock.Lock()
-				defer lock.Unlock()
+				testCase.lock.Lock()
+				defer testCase.lock.Unlock()
 
 				pod := testCase.pod.(*v1.Pod)
 				pod.Spec.Containers[0].Name = "image2"
@@ -352,9 +353,7 @@ func watchFileChanged(watchDir bool, symlink bool, t *testing.T) {
 			expectUpdate(t, ch, testCase)
 
 			if watchDir {
-				from := fileName
-				fileName = fileName + "_ch"
-				go changeFileName(dirName, from, fileName, t)
+				go changeFileName(dirName, fileName, fileName+"_ch", t)
 				// expect an update by MOVED_FROM inotify event cause changing file name
 				expectEmptyUpdate(t, ch)
 				// expect an update by MOVED_TO inotify event cause changing file name
@@ -365,8 +364,12 @@ func watchFileChanged(watchDir bool, symlink bool, t *testing.T) {
 }
 
 func deleteFile(dir, file string, ch chan interface{}, t *testing.T) {
+	path := filepath.Join(dir, file)
+	if _, err := os.Stat(path); err != nil {
+		// The file might absent if it's renamed, deleted..
+		return
+	}
 	go func() {
-		path := filepath.Join(dir, file)
 		err := os.Remove(path)
 		if err != nil {
 			t.Errorf("unable to remove test file %s: %s", path, err)
@@ -397,6 +400,8 @@ func expectUpdate(t *testing.T, ch chan interface{}, testCase *testCase) {
 				}
 			}
 
+			testCase.lock.Lock()
+			defer testCase.lock.Unlock()
 			if !apiequality.Semantic.DeepEqual(testCase.expected, update) {
 				t.Fatalf("%s: Expected: %#v, Got: %#v", testCase.desc, testCase.expected, update)
 			}

